### PR TITLE
feat(multisig-prover): encode bcs proof

### PR DESCRIPTION
**Pull request by cjcobb23**
https://axelarnetwork.atlassian.net/browse/AXE-1912

### CODE REVIEW

File: contracts/multisig-prover/src/encoding/bcs.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+use multisig::{key::Signature, msg::Signer};
+
+use crate::{error::ContractError, types::Operator};
+
+#[allow(dead_code)]
+fn encode_proof(
+    quorum: Uint256,
+    signers: Vec<(Signer, Option<Signature>)>,
+) -> Result<HexBinary, ContractError> {
+    let mut operators = make_operators_with_sigs(signers)?;
```
Comment (sammy1991106) [inline contracts/multisig-prover/src/encoding/bcs.rs:13]: sorting is a requirement due to our implementation of the EVM gateway contract. Does the sui contract also require the same?

File: contracts/multisig-prover/src/encoding/bcs.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            .try_into()
+            .expect("couldn't convert u256 to u128"),
+    );
+    Ok(to_bytes(&(addresses, weights, quorum, signatures))?.into())
+}
+
+#[allow(dead_code)]
+fn make_operators_with_sigs(
+    signers_with_sigs: Vec<(Signer, Option<Signature>)>,
+) -> Result<Vec<Operator>, ContractError> {
```
Comment (sammy1991106) [inline contracts/multisig-prover/src/encoding/bcs.rs:44]:
```suggestion
) -> Vec<Operator> {
```

File: contracts/multisig-prover/src/encoding/bcs.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    let signatures: Vec<Vec<u8>> = signatures.into_iter().flatten().collect();
+    let quorum = &u128::from_le_bytes(
+        quorum.to_le_bytes()[..16]
+            .try_into()
+            .expect("couldn't convert u256 to u128"),
+    );
+    Ok(to_bytes(&(addresses, weights, quorum, signatures))?.into())
+}
+
+#[allow(dead_code)]
```
Comment (sammy1991106) [inline contracts/multisig-prover/src/encoding/bcs.rs:41]:
```suggestion
```
Don't need to put a on a private fn

File: contracts/multisig-prover/src/encoding/bcs.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                )
+                .to_le(),
+                op.signature.as_ref().map(|sig| sig.as_ref().to_vec()),
+            )
+        })
+        .multiunzip();
+
+    let signatures: Vec<Vec<u8>> = signatures.into_iter().flatten().collect();
+    let quorum = &u128::from_le_bytes(
+        quorum.to_le_bytes()[..16]
```
Comment (sammy1991106) [inline contracts/multisig-prover/src/encoding/bcs.rs:34]: smae here. Should validate the length?

File: contracts/multisig-prover/src/encoding/bcs.rs (hunk 1 above)
Comment (cjcobb23) [reply contracts/multisig-prover/src/encoding/bcs.rs:13]: yeah sui requires sorting as well

File: contracts/multisig-prover/src/encoding/bcs.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+#[allow(dead_code)]
+fn encode_proof(
+    quorum: Uint256,
+    signers: Vec<(Signer, Option<Signature>)>,
+) -> Result<HexBinary, ContractError> {
+    let mut operators = make_operators_with_sigs(signers);
+    operators.sort();
+
+    let (addresses, weights, signatures): (Vec<_>, Vec<u128>, Vec<_>) = operators
```
Comment (cgorenflo) [inline contracts/multisig-prover/src/encoding/bcs.rs:29]: do you need `u128` here? Is the signature of `u256_to_u128` not enough for the compiler?

**Review by cgorenflo — APPROVED** [review]

File: contracts/multisig-prover/src/encoding/bcs.rs (hunk 1 above)
Comment (cgorenflo) [reply contracts/multisig-prover/src/encoding/bcs.rs]: document that on the code?

**Review by sammy1991106 — APPROVED** [review]

**Pull request merged**